### PR TITLE
Refactor the asynchronous process_spider_output documentation

### PR DESCRIPTION
**Pull request by Gallaecio**
I am suggesting a refactoring of the documentation.

These are some of the ideas behind this suggestion:

-   I would avoid future references as much as possible where user action is
    not needed, i.e. do not mention that this will apply to
    ``process_start_requests`` in the future, and write the documentation with
    the current situation in mind only (talking about “these methods” when it
    only applies to 1 method at the time can be confusing; I don’t think being
    able to avoid editing some lines of documentation in the future is worth
    it).

-   I would stick to the vocabulary as defined in the Python glossary to
    minimize confusion, and include links to the glossary in the first
    reference in any section.

-   I take it “they can take a normal one and return an async one” was a
    remnant from a previous iteration, before the current universal approach,
    and I have removed that.

-   I have aimed at removing references to what Scrapy does except for the
    specific scenario where user action is needed (asynchronous-to-synchronous
    conversion), instead focusing on what the user will do and get (how they
    need to define their methods, and expected input parameter types).

Pending aspects that I want to test or research before this pull request can be
reviewed:

-   [x] What happens in the case of a coroutine that returns an iterable? Would
    the user get a helpful error message? Otherwise, should we conver something
    here or in the FAQ to help users figure out what they did wrong?

    >   I am suggesting a more specific error message for this scenario, and I
    >   think that is enough, i.e. no need to cover this specifically in the
    >   documentation, where we already specifically indicate that an
    >   asynchronous generator must be used.

-   [x] Do async request callbacks need to be asynchronous generators (i.e..
    use yield) or can they return values instead? How does our inspect code
    behave if they combine yield and return?

    >   They can be regular coroutines, and support the same output values as
    >   synchronous functions.
    >
    >   Our code that checks that yield is not combined with a valued return
    >   is irrelevant here because doing that in a coroutine causes a syntax
    >   error altogether.

-   [x] When an exception happens as part of an async-to-sync conversion, what
    happens exactly? Are all items skipped always? Are items processed before
    the exception still passed forward? How do things behave in similar
    scenarios not involving a conversion, either all sync or all async? What
    is the behavior on sync-to-async conversions?

    >   Since this is not explained in the regular spider middleware
    >   documentation, as far as I can tell, I have extended the corresponding
    >   point of the new documentation to clarify that, when there is no
    >   async-to-sync conversion, items yielded before an exception is raised
    >   are processed.
    >
    >   Also, this scenario seems covered by tests at
    >   `tests/test_spidermiddleware_output_chain.py`.

-   [x] The current documentation says:

    >   Async iterables returned from
    >   :meth:`~scrapy.spidermiddlewares.SpiderMiddleware.process_spider_exception`
    >   won't be downgraded, an exception will be raised if that is needed.

    What does that mean, exactly? Can process_spider_exception be defined as a
    coroutine?

    >   We decided not to document coroutine support for this method for now.
    >
    >   Later on we will figure out what to do with its coroutine API.

-   [x] Restore the tip about using ``scrapy.__version__`` when writing
    async-only middlewares. Maybe we could give it its own section, and include
    an example. And maybe we could write an asyncio counterpart section in the
    asyncio page, featuring gracefully disabling Scrapy components if an
    asyncio reactor is not being used, also featuring an example.
    
    >   Done through a new documentation page that may also be used for
    >   https://github.com/scrapy/scrapy/issues/5110

### CODE REVIEW

File: docs/topics/coroutines.rst [hunk 1]
```diff
@@ -34,18 +37,26 @@ hence use coroutine syntax (e.g. ``await``, ``async for``, ``async with``):
...
+=============
+
+There are several use cases for coroutines in Scrapy.
 
-There are several use cases for coroutines in Scrapy. Code that would
-return Deferreds when written for previous Scrapy versions, such as downloader
-middlewares and signal handlers, can be rewritten to be shorter and cleaner::
+Code that would return Deferreds when written for previous Scrapy versions,
+such as downloader middlewares and signal handlers, can be rewritten to be
+shorter and cleaner::
```
Comment (Gallaecio) [inline docs/topics/coroutines.rst:59]: Since we are adding additional usage sections, I thought it better to name this one “General usage” instead.

I also took the chance to split the first paragraph (unrelated change).

File: docs/topics/asyncio.rst [hunk 2]
```diff
@@ -64,6 +63,9 @@ yourself, or in some code that runs before the reactor is installed, e.g.
 
 .. _playwright: https://github.com/microsoft/playwright-python
 
+
+.. _asyncio-await-dfd:
```
Comment (Gallaecio) [inline docs/topics/asyncio.rst]: Unrelated documentation fix.

File: scrapy/core/spidermw.py [hunk 3]
```diff
@@ -61,7 +61,7 @@ def _process_spider_input(self, scrape_func: ScrapeFunc, response: Response, req
             try:
                 result = method(response=response, spider=spider)
                 if result is not None:
-                    msg = (f"Middleware {method.__qualname__} must return None "
```
Comment (Gallaecio) [inline scrapy/core/spidermw.py:64]: Since it is a middleware method, not a middleware, “Method” or “Middleware method” would be the right choice instead of “Middleware”. However, I don’t think we need a word at all, we are already providing the qualified path.

File: docs/topics/coroutines.rst [hunk 4]
```diff
@@ -110,46 +121,76 @@ Common use cases for asynchronous code include:
...
+callbacks as coroutine methods or when using spider middlewares whose
+``process_spider_output`` method is an :term:`asynchronous generator`, all
+active spider middlewares must either have their ``process_spider_output``
+method defined as an asynchronous generator or :ref:`define a
+process_spider_output_async method <universal-spider-middleware>`.
+
+.. note:: When using third-party spider middlewares that only define a
+          synchronous ``process_spider_output`` method, consider
+          :ref:`making them universal <universal-spider-middleware>` through
+          :ref:`subclassing <tut-inheritance>`.
```
Comment (kmike) [inline docs/topics/coroutines.rst:175]: If you're subclassing a third party middleware for use in your project, why does it need to be universal, i.e. why does it need to support older Scrapy versions?

**Review by kmike — APPROVED** [review]
Looks good to me, thanks @Gallaecio!

File: docs/topics/coroutines.rst (hunk 4 above)
Comment (Gallaecio) [reply docs/topics/coroutines.rst:175]: If I recall right, the reasoning is that it is the easiest way for users.

For example, if the third-party class is:

```python
class ThirdPartyMiddleware:
    def process_spider_output(self, response, result, spider):
        ...
```

A universal subclass would always look like this (untested):

```python
class UniversalThirdPartyMiddleware(ThirdPartyMiddleware):
    async def process_spider_output_async(self, response, result, spider):
        yield from self.process_spider_output(response, result, spider)
```

Assuming I got it right (cc @wRAR), maybe we could be more explicit about that here, and include an example like the one in this comment to make things clear.

We could probably create some utility to allow creating a universal subclass with a single line, come to think of it…

**Comment by Gallaecio** [comment]
I just had a thought: since `async def` will work for all callbacks, should we promote its use in the docs? (i.e. use it anywhere where a callback is defined in the docs)

**Comment by kmike** [comment]
> I just had a thought: since async def will work for all callbacks, should we promote its use in the docs? (i.e. use it anywhere where a callback is defined in the docs)

What do you think about making this change later, after getting more experience with async def parse methods?

File: docs/topics/coroutines.rst (hunk 4 above)
Comment (wRAR) [reply docs/topics/coroutines.rst:175]: I'm fine with adding this example (though maybe we should test it first), not sure about creating a helper.

**Review by wRAR — APPROVED** [review]
